### ISSUE TITLE
tfcompile with negative one shape

### ISSUE THREAD

**Issue by lotrus28** [tensorflow/tensorflow#32161]
I am trying to compile a tensorflow model, using the following [tutorial](https://barkeywolf.consulting/posts/tf-aot-rust/#prepare-the-tensorflow-graph-for-compilation), which is based on [tf documentation](https://www.tensorflow.org/xla/tfcompile).
<br>
I am doing this in the latest `tensorflow/tensorflow` docker.
<br>
All the tutorials I have seen deal with image processing, but my models receive one-dimensional vector. The config file looks like this:

```
feed {
  id {
    node_name: "dense_1_input"
  }
  shape {
    dim {
      size: -1
    }
    dim {
      size: 5
    }
  }
}
fetch {
  id {
    node_name: "k2tfout_0"
  }
}
```

If I use it, the compilation fails. If I change `-1` for `1`, the model compiles, but whenever I try to use it, I get a core dump.
<br>
Could this be an issue with this negative one dimension? How do I compile such models?
<br>
I tried invoking my model using Python and flipping dimension numbers. (1,5) – (5,1) – (5,)...
<br>E.g.:

```
import numpy as np
import pandas as pd

libmodel = np.ctypeslib.load_library('libmodel', '/folder/org_tensorflow/')

libmodel.run.argtypes = [np.ctypeslib.ndpointer(np.float32, ndim=2, shape=(1,5), flags=('c', 'a')),
np.ctypeslib.ndpointer(np.float32, ndim=2, shape=(1,5), flags=('c', 'a', 'w')),
np.ctypeslib.ctypes.c_float,
np.ctypeslib.ctypes.c_float]

df = pd.read_csv('/folder/trial.csv', header = 0, index_col = None)
feats = pd.read_table('/folder/feature_list.txt', header=None, index_col = None)
feats = feats.iloc[:,0].tolist()
df = df.reindex(columns=feats, fill_value=0)

x = np.require(df.iloc[0,:], np.float32, ('c', 'a'))
y = np.require(np.zeros((1,1)), np.float32, ('c', 'a', 'w'))

x = x.reshape((1,5))
libmodel.run(x, y, x.size, y.size)
```

Is it possible to compile such models with negative one dimension? If so, it should be in documentation.

**Comment by gadagashwini** [comment]
@lotrus28, In order to expedite the trouble-shooting process, please provide a complete code snippet to reproduce the issue reported here. Thanks!

**Comment by lotrus28** [comment]
> @lotrus28, In order to expedite the trouble-shooting process, please provide a complete code snippet to reproduce the issue reported here. Thanks!

Here is the docker file, which is basically 

```
FROM tensorflow/tensorflow:latest-devel-py3

RUN git clone --branch=master --no-progress https://github.com/tensorflow/tensorflow.git /tensorflow

WORKDIR /tensorflow
ENV TF_NEED_JEMALLOC=1 \
    TF_NEED_GCP=0 \
    TF_NEED_HDFS=0 \
    TF_NEED_AWS=0 \
    TF_NEED_KAFKA=0 \
    TF_ENABLE_XLA=0 \
    TF_NEED_GDR=0 \
    TF_NEED_VERBS=0 \
    TF_NEED_OPENCL_SYCL=0 \
    TF_NEED_OPENCL=0 \
    TF_NEED_CUDA=0 \
    TF_CUDA_CLANG=0 \
    TF_NEED_ROCM=0 \
    TF_NEED_MKL=0 \
    TF_DOWNLOAD_MKL=0 \
    TF_DOWNLOAD_CLANG=0 \
    TF_NEED_MPI=0 \
    TF_NEED_S3=0 \
    TF_SET_ANDROID_WORKSPACE=0 \
    TF_NEED_COMPUTECPP=0 \
    TF_NEED_TENSORRT=0 \
    TF_CONFIGURE_IOS=0 \    
    CC_OPT_FLAGS="-march=native -Wno-sign-compare" 
RUN export PYTHON_BIN_PATH=$(which python3);\
    export PYTHON_LIB_PATH="$($PYTHON_BIN_PATH -c 'import site; print(site.getsitepackages()[0])')";\
    python3 configure.py
RUN bazel build -c opt --verbose_failures //tensorflow/compiler/aot:tfcompile
RUN	exit 0
```

Then I copy these files to the tensorflow directory:

> BUILD
```

load('@org_tensorflow//tensorflow/compiler/aot:tfcompile.bzl', 'tf_library')

tf_library(
    name = 'graph',
    config = 'graph.config.pbtxt',
    cpp_class = 'Graph',
    graph = 'graph.pb',
)

cc_binary(
    name = "libmodel.so",
    srcs = ["graph.cc"],
    deps = [":graph", "//third_party/eigen3"],
    linkopts = ["-lpthread"],
    linkshared = 1,
    copts = ["-fPIC"],
)
```

> graph.cc
```
#define EIGEN_USE_THREADS
#define EIGEN_USE_CUSTOM_THREAD_POOL

#include "graph.h"
#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"

extern "C" int run(
		float *input,
		float *output_age,
		int input_size,
		int output_age_size
    ){
	// allocate an instance of the Graph, along with all its private buffers
	Eigen::ThreadPool tp(std::thread::hardware_concurrency());
	Eigen::ThreadPoolDevice device(&tp, tp.NumThreads());
	Graph graph;
	graph.set_thread_pool(&device);

	// copy over the input buffer
	std::copy(input, input + input_size, graph.arg0_data());

	// execute the inference graph
	auto ok = graph.Run();
	if (not ok) return -1;

	// copy the result into the output buffer
	std::copy(graph.result0_data(), graph.result0_data() + output_age_size, output_age);
	return 0;
}
```

I have already shown you my `graph.config.pbtxt`. 

To start compiling I do 
`bazel build --show_progress_rate_limit=60 @org_tensorflow//:libmodel.so`

If I check the model's shape in Python:
```
>>m.input_shape
>> (None, 5)
```
But saving it as `.pb` produces a `-1` dimension in the human readable layers' descriptions.

**Comment by lotrus28** [comment]
OK, seems like tensorflow models uses `-1` for a dynamic number of inputs.
E.g. my network can receive 1 or 10 or 100 5-long vectors. And it is likely I cannot compile a model with a dynamic input.<br>
Unfortunately I am unable to check if the model I compiled will work when I enter only one 5-long vector to it. Hopefully, it will.<br>
Does that sound reasonable?

**Comment by lotrus28** [comment]
Seems like `-1` in the inputs stands for a dynamic input. I.e. my model can receive 1 or 10 or 100 5-long vectors to produce an output. When I specify shape as (1,5) in the config, it compiles. Does that mean that the compiled model expects (1,5) input? Well, I do just that, when invoking:<br>
`x = np.require(df.iloc[0,:], np.float32, ('c', 'a'))`<br>
Could the problem be in the way I specify the output? I also tried `dim = 1`, but that did not work either.

**Comment by lotrus28** [comment]
Ok, I tried to totally reduce the model completely and try to compile it without training. This time I defined the input shape explicitly, so that its input shape is `(1, 2)` in contrast to `(None, 2)`, if I defined it the way I used to.
<br>
```
>> model.add(Dense(5, input_shape=(2,), batch_size = 1, kernel_initializer='uniform'))
>> model.add(BatchNormalization())
>> model.add(PReLU())
>> model.add(Dropout(0.5))
>> model.add(Dense(1, kernel_initializer='normal'))
>> model.compile(loss='mean_absolute_error', optimizer='adam')
>> model.save('../test.h5')
```
The summary is then:<br>
```
Layer (type)                 Output Shape              Param #   
=================================================================
dense_1 (Dense)              (1, 5)                    15        
_________________________________________________________________
batch_normalization_1 (Batch (1, 5)                    20        
_________________________________________________________________
p_re_lu_1 (PReLU)            (1, 5)                    5         
_________________________________________________________________
dropout_1 (Dropout)          (1, 5)                    0         
_________________________________________________________________
dense_2 (Dense)              (1, 1)                    6         
=================================================================
Total params: 46
Trainable params: 36
Non-trainable params: 10
```
`keras_to_tensorflow.py` identifies its output as `dense_2/BiasAdd`
Is it possible to compile this model?  If not, what modifications can I do to make it possible, while preserving the weights?

**Comment by lotrus28** [comment]
I have attached the compilation log for a model with just 2 Dense layers.
As before, it compiles, but can't be further used. Perhaps there are some important warnings in there that could help

[LOG.txt](https://github.com/tensorflow/tensorflow/files/3579259/LOG.txt)

**Comment by jvishnuvardhan** [comment]
@lotrus28 What do you mean by the following. 

> > it compiles, but can't be further used.

 
Here is the [gist](https://colab.sandbox.google.com/gist/jvishnuvardhan/1665ec9a6ba00b2e001f8c37cf86eb54/tf32161.ipynb). Can you check my gist and add the part that is not working? Thanks!

**Comment by lotrus28** [comment]
> 
> 
> @lotrus28 What do you mean by the following.
> 
> > > it compiles, but can't be further used.
> 
I mean that if I use `tfcompile` to convert a network into binary, I do get an `.so` file, but during invocation I get core dump error.<br>
The invocation fails even if I start with a model containing only 2 dense layers.
I tried fixing the batch dimension and feeding and reshaping the input, but I still get a segmentation error. So, most likely the error is in this part

```
x = np.require(np.zeros((1,2)), np.float32, ('c', 'a'))
y = np.require(np.zeros((1,1)), np.float32, ('c', 'a', 'w'))
res = libmodel.run(x, y, x.size, y.size)
```

How would you run this sample network?

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
